compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -5169,21 +5170,20 @@ object Types extends TypeUtils {
       then
         record("MatchType.reduce computed")
         if (myReduced != null) record("MatchType.reduce cache miss")
-        myReduced =
-          trace(i"reduce match type $this $hashCode", matchTypes, show = true):
+        if !isUpToDate then setReductionContext()
+        val saved = ctx.typerState.snapshot()
+        try
+          myReduced = trace(i"reduce match type $this $hashCode", matchTypes, show = true):
```

Comment:
I see now: We did cache reduce failures before _except those that threw an exception_. That's handled now.